include/swift/SIL/SILSuccessor.h

```diff
@@ -87,33 +87,45 @@ class SILSuccessor {
   class pred_iterator {
     SILSuccessor *Cur;
 
+    // Cache the basic block to avoid repeated pointer chasing.
+    SILBasicBlock *BB;
+
+    void cacheBasicBlock();
+
   public:
     using difference_type = std::ptrdiff_t;
     using value_type = SILBasicBlock *;
     using pointer = SILBasicBlock **;
     using reference = SILBasicBlock *&;
     using iterator_category = std::forward_iterator_tag;
 
-    pred_iterator(SILSuccessor *Cur = 0) : Cur(Cur) {}
+    pred_iterator(SILSuccessor *Cur = nullptr) : Cur(Cur), BB(nullptr) {
+      cacheBasicBlock();
+    }
 
     bool operator==(pred_iterator I2) const { return Cur == I2.Cur; }
     bool operator!=(pred_iterator I2) const { return Cur != I2.Cur; }
 
     pred_iterator &operator++() {
-      assert(Cur && "Trying to advance past end");
+      // No need to assert(Cur). Dereferencing nullptr always crashes.
```

Comment:
We want a nice error message here. I would suggest actually making Cur a NullablePtr to make it explicit.

Reply:
I never understood the "crash intentionally (a.k.a. "assert") before an implicit crash" convention. What problem is this solving?

Reply:
If you are looking at why something crashed, it provides a really nice error message without you having to load up the debugger.